pkg/template/controller/templateinstance_controller.go

```diff
@@ -0,0 +1,174 @@
+	"k8s.io/kubernetes/pkg/api/meta/metatypes"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/auth/user"
+	"k8s.io/kubernetes/pkg/client/cache"
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type TemplateInstanceController struct {
+	restconfig restclient.Config
+	oc         *client.Client
+	controller *cache.Controller
+}
+
+func NewTemplateInstanceController(restconfig restclient.Config, oc *client.Client) *TemplateInstanceController {
+	c := TemplateInstanceController{restconfig: restconfig, oc: oc}
+
+	_, c.controller = cache.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
+				return c.oc.TemplateInstances(kapi.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
+				return c.oc.TemplateInstances(kapi.NamespaceAll).Watch(options)
+			},
+		},
+		&templateapi.TemplateInstance{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.handle(obj.(*templateapi.TemplateInstance))
+			},
+			UpdateFunc: func(_, obj interface{}) {
+				c.handle(obj.(*templateapi.TemplateInstance))
+			},
+			DeleteFunc: func(obj interface{}) {
+			},
+		},
+	)
+
+	return &c
+}
+
+func (c *TemplateInstanceController) Run(stop <-chan struct{}) {
+	c.controller.Run(stop)
+}
+
+func (c *TemplateInstanceController) handle(templateInstance *templateapi.TemplateInstance) error {
+	for _, condition := range templateInstance.Status.Conditions {
+		if condition.Type == templateapi.TemplateInstanceCreated && condition.Status == kapi.ConditionTrue ||
+			condition.Type == templateapi.TemplateInstanceFailed && condition.Status == kapi.ConditionTrue {
+			return nil
+		}
+	}
+
+	err := c.provision(templateInstance)
+	if err == nil {
+		templateInstance.Status.Conditions = []templateapi.TemplateInstanceCondition{
+			{
+				Type:               templateapi.TemplateInstanceCreated,
+				Status:             kapi.ConditionTrue,
+				LastTransitionTime: unversioned.Now(),
+				Reason:             "Created",
+			},
+		}
+
+	} else {
+		templateInstance.Status.Conditions = []templateapi.TemplateInstanceCondition{
+			{
+				Type:               templateapi.TemplateInstanceFailed,
+				Status:             kapi.ConditionTrue,
+				LastTransitionTime: unversioned.Now(),
+				Reason:             "Failed",
+				Message:            err.Error(),
+			},
+		}
+	}
+
+	_, err = c.oc.TemplateInstances(templateInstance.Namespace).Update(templateInstance)
+	if err != nil {
+		glog.Errorf("TemplateInstance update failed: %v", err)
+	}
+	return err
+}
+
+func (c *TemplateInstanceController) provision(templateInstance *templateapi.TemplateInstance) error {
+	config := c.restconfig
+	config.WrapTransport = func(rt http.RoundTripper) http.RoundTripper {
```

Comment:
I would have assumed David had already abstracted this somewhere in the code (probably project request or the jenkins admission plugin) to make this possible.  Would prefer all three places either share, or start pulling shared code out with the intent to merge them all under one common pattern (instantiateTemplateAs(...))